File: src/utils/errorHandler.js
/**
 * Error Handler
 * Global error handling and reporting
 */

const { Logger } = require('./logger');

class ErrorHandler {
  constructor(client) {
    this.client = client;
    this.logger = Logger.create('ErrorHandler');
    this.errorCounts = new Map();
    this.maxErrorsPerMinute = 10;
    this.circuitBreakers = new Map();
  }

  /**
   * Initialize global error handlers
   */
  initialize() {
    // Handle uncaught exceptions
    process.on('uncaughtException', (error) => {
      this.handleFatalError('Uncaught Exception', error);
    });

    // Handle unhandled promise rejections
    process.on('unhandledRejection', (reason, promise) => {
      this.handleFatalError('Unhandled Rejection', reason);
    });

    // Handle Discord.js errors
    this.client.on('error', (error) => {
      this.logger.error(`Discord client error: ${error.message}`);
    });

    this.client.on('shardError', (error, shardId) => {
      this.logger.error(`Shard ${shardId} error: ${error.message}`);
    });

    // Handle disconnections
    this.client.on('disconnect', (event) => {
      this.logger.warn(`Client disconnected: ${event.code} - ${event.reason}`);
    });

    // Periodic cleanup of error counts
    setInterval(() => this.cleanupErrorCounts(), 60000);

    this.logger.info('Error handlers initialized');
  }

  /**
   * Handle a fatal error (may crash the process)
   * @param {string} type - Error type
   * @param {Error} error - Error object
   */
  handleFatalError(type, error) {
    this.logger.error(`=== FATAL ERROR: ${type} ===`);
    this.logger.error(error.stack || error.message || error);

    // Log system state
    this.logSystemState();

    // Graceful shutdown attempt
    this.gracefulShutdown();
  }

  /**
   * Handle operational errors (recoverable)
   * @param {string} context - Error context
   * @param {Error} error - Error object
   * @returns {boolean} Whether error count exceeded threshold
   */
  handleOperationalError(context, error) {
    const key = `${context}:${error.message}`;
    const count = (this.errorCounts.get(key) || 0) + 1;
    this.errorCounts.set(key, count);

    // Log with context
    this.logger.error(`[${context}] ${error.message}`, {
      count,
      stack: error.stack?.split('\n')[0],
    });

    // Check if we should circuit break
    if (count >= this.maxErrorsPerMinute) {
      this.logger.warn(`Circuit breaker triggered for: ${context}`);
      this.circuitBreakers.set(context, Date.now() + 60000); // Block for 1 minute
      return true;
    }

    return false;
  }

  /**
   * Check if a context is circuit broken
   * @param {string} context - Context to check
   * @returns {boolean}
   */
  isCircuitBroken(context) {
    const breakUntil = this.circuitBreakers.get(context);
    if (!breakUntil) return false;

    if (Date.now() > breakUntil) {
      this.circuitBreakers.delete(context);
      this.errorCounts.delete(context);
      return false;
    }

    return true;
  }

  /**
   * Wrap an async function with error handling
   * @param {string} context - Error context
   * @param {Function} fn - Function to wrap
   * @returns {Function}
   */
  wrap(context, fn) {
    return async (...args) => {
      if (this.isCircuitBroken(context)) {
        throw new Error(`Circuit breaker active for: ${context}`);
      }

      try {
        return await fn(...args);
      } catch (error) {
        this.handleOperationalError(context, error);
        throw error;
      }
    };
  }

  /**
   * Log current system state
   */
  logSystemState() {
    const usage = process.memoryUsage();
    this.logger.info('System State:', {
      memory: {
        rss: `${Math.round(usage.rss / 1024 / 1024)}MB`,
        heapUsed: `${Math.round(usage.heapUsed / 1024 / 1024)}MB`,
        heapTotal: `${Math.round(usage.heapTotal / 1024 / 1024)}MB`,
      },
      uptime: `${Math.round(process.uptime())}s`,
      errors: this.errorCounts.size,
    });
  }

  /**
   * Get error statistics
   * @returns {Object}
   */
  getStats() {
    const stats = {
      totalErrors: 0,
      byContext: {},
      circuitBreakers: Array.from(this.circuitBreakers.keys()),
    };

    for (const [key, count] of this.errorCounts) {
      stats.totalErrors += count;
      stats.byContext[key] = count;
    }

    return stats;
  }

  /**
   * Cleanup old error counts
   */
  cleanupErrorCounts() {
    const hadErrors = this.errorCounts.size > 0;
    this.errorCounts.clear();
    // Only log if there were errors to clean up
    if (hadErrors) {
      this.logger.debug('Error counts cleaned up');
    }
  }

  /**
   * Graceful shutdown attempt
   */
  gracefulShutdown() {
    this.logger.info('Attempting graceful shutdown...');

    // Give logger time to flush
    setTimeout(() => {
      process.exit(1);
    }, 1000);
  }
}

module.exports = { ErrorHandler };

File: utils/error_handler.py
"""
Error Handler
Global error handling and reporting
"""

import asyncio
import signal
import sys
import traceback
from typing import Any, Callable, Dict, Optional

from utils.logger import get_logger

logger = get_logger("ErrorHandler")


class ErrorHandler:
    """Global error handler for the bot."""

    def __init__(self):
        self.logger = get_logger("ErrorHandler")
        self.error_counts: Dict[str, int] = {}
        self.circuit_breakers: Dict[str, float] = {}
        self.max_errors_per_minute = 10
        self._cleanup_task: Optional[asyncio.Task] = None

    def initialize(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
        """Initialize global error handlers."""
        if loop is None:
            loop = asyncio.get_event_loop()

        # Set up exception handlers
        loop.set_exception_handler(self._async_exception_handler)

        # Handle signals
        try:
            loop.add_signal_handler(signal.SIGINT, self._signal_handler, signal.SIGINT)
            loop.add_signal_handler(signal.SIGTERM, self._signal_handler, signal.SIGTERM)
        except NotImplementedError:
            # Windows doesn't support add_signal_handler
            pass

        # Start periodic cleanup
        self._cleanup_task = asyncio.create_task(self._periodic_cleanup())

        self.logger.info("Error handlers initialized")

    def _signal_handler(self, sig: signal.Signals) -> None:
        """Handle shutdown signals."""
        self.logger.info(f"Received signal {sig.name}, shutting down...")
        asyncio.create_task(self.shutdown())

    def _async_exception_handler(self, loop: asyncio.AbstractEventLoop, context: Dict[str, Any]) -> None:
        """Handle async exceptions."""
        exception = context.get("exception")
        if exception:
            self.handle_exception(exception)
        else:
            message = context.get("message", "Unknown async error")
            self.logger.error(f"Async error: {message}")

    def handle_exception(self, error: Exception, context: str = "") -> bool:
        """
        Handle an exception.

        Args:
            error: The exception that occurred
            context: Optional context string

        Returns:
            True if error count exceeded threshold (circuit broken)
        """
        error_key = f"{context}:{type(error).__name__}"

        # Log the error
        if context:
            self.logger.error(f"[{context}] {error}")
        else:
            self.logger.error(f"{error}")

        # Log traceback for debugging
        self.logger.debug(f"Traceback:\n{traceback.format_exc()}")

        # Update error count
        count = self.error_counts.get(error_key, 0) + 1
        self.error_counts[error_key] = count

        # Check if we should circuit break
        if count >= self.max_errors_per_minute:
            self.logger.warning(f"Circuit breaker triggered for: {error_key}")
            self.circuit_breakers[error_key] = asyncio.get_event_loop().time() + 60  # Block for 1 minute
            return True

        return False

    def is_circuit_broken(self, context: str) -> bool:
        """Check if a context is circuit broken."""
        break_until = self.circuit_breakers.get(context)
        if not break_until:
            return False

        if asyncio.get_event_loop().time() > break_until:
            # Circuit breaker expired
            del self.circuit_breakers[context]
            self.error_counts.pop(context, None)
            return False

        return True

    async def _periodic_cleanup(self) -> None:
        """Periodically clean up error counts."""
        while True:
            try:
                await asyncio.sleep(60)  # Clean up every minute
                self._cleanup_error_counts()
            except asyncio.CancelledError:
                break
            except Exception as e:
                self.logger.debug(f"Cleanup error: {e}")

    def _cleanup_error_counts(self) -> None:
        """Clean up old error counts."""
        if self.error_counts:
            self.logger.debug("Error counts cleaned up")
        self.error_counts.clear()

    def wrap(self, context: str):
        """Decorator to wrap a function with error handling."""
        def decorator(func: Callable) -> Callable:
            async def wrapper(*args, **kwargs):
                if self.is_circuit_broken(context):
                    raise RuntimeError(f"Circuit breaker active for: {context}")

                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    self.handle_exception(e, context)
                    raise
            return wrapper
        return decorator

    async def shutdown(self) -> None:
        """Clean shutdown."""
        self.logger.info("Shutting down error handler...")

        if self._cleanup_task:
            self._cleanup_task.cancel()
            try:
                await self._cleanup_task
            except asyncio.CancelledError:
                pass

        # Clear error counts
        self.error_counts.clear()
        self.circuit_breakers.clear()

    def log_system_state(self) -> None:
        """Log current system state."""
        import psutil

        process = psutil.Process()
        memory = process.memory_info()

        self.logger.info(
            "System state: "
            f"memory={memory.rss / 1024 / 1024:.1f}MB, "
            f"errors={len(self.error_counts)}, "
            f"circuits={len(self.circuit_breakers)}"
        )


# Global error handler instance
_error_handler: Optional[ErrorHandler] = None


def get_error_handler() -> ErrorHandler:
    """Get the global error handler instance."""
    global _error_handler
    if _error_handler is None:
        _error_handler = ErrorHandler()
    return _error_handler


def setup_error_handler(loop: Optional[asyncio.AbstractEventLoop] = None) -> ErrorHandler:
    """Set up the global error handler."""
    handler = get_error_handler()
    handler.initialize(loop)
    return handler
